packages/cdktf-cli/lib/cdktf-project.ts

```diff
@@ -519,51 +524,47 @@ export class CdktfProject {
   private async execute(
     method: "deploy" | "destroy",
     next: () => Promise<CdktfStack | undefined>,
-    parallelism: number,
-    refreshOnly?: boolean
+    opts: MutationOptions
   ) {
     // We only support refresh only on deploy, a bit of a leaky abstraction here
-    if (refreshOnly && method !== "deploy") {
+    if (opts.refreshOnly && method !== "deploy") {
       throw Errors.Internal(`Refresh only is only supported on deploy`);
     }
 
-    const maxParallelRuns = parallelism === -1 ? Infinity : parallelism;
+    const maxParallelRuns =
+      !opts.parallelism || opts.parallelism < 0 ? Infinity : opts.parallelism;
     while (this.stacksToRun.filter((stack) => stack.isPending).length > 0) {
       const runningStacks = this.stacksToRun.filter((stack) => stack.isRunning);
       if (runningStacks.length >= maxParallelRuns) {
         await Promise.race(runningStacks.map((s) => s.currentWorkPromise));
-      } else {
-        try {
-          const nextRunningExecutor = await next();
-          if (!nextRunningExecutor) {
-            // In this case we have no pending stacks, but we also can not find a new executor
-            break;
-          }
-          method === "deploy"
-            ? nextRunningExecutor.deploy(refreshOnly)
-            : nextRunningExecutor.destroy();
-        } catch (e) {
-          // await next() threw an error because a stack failed to apply/destroy
-          // wait for all other currently running stacks to complete before propagating that error
-          logger.debug(
-            "Encountered an error while awaiting stack to finish",
-            e
-          );
-          const openStacks = this.stacksToRun.filter(
-            (ex) => ex.currentWorkPromise
-          );
-          logger.debug(
-            "Waiting for still running stacks to finish:",
-            openStacks
-          );
-          await Promise.allSettled(
-            openStacks.map((ex) => ex.currentWorkPromise)
-          );
-          logger.debug(
-            "Done waiting for still running stacks. All pending work finished"
-          );
-          throw e;
+        continue;
```

Comment:
Flattening the if-else a bit. As far as I can see, it shouldn't have any impact, but maybe I've missed something